core/serialization/serialization.js

```diff
@@ -0,0 +1,87 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Contains top-level functions for serialization of the workspace
+ *     to JavaScript objects.
+ */
+'use strict';
+
+/**
+ * The top level namespace for JavaScript Object serialization.
+ * @namespace Blockly.serialization
+ */
+goog.module('Blockly.serialization');
+goog.module.declareLegacyNamespace();
+ 
+const blocks = goog.require('Blockly.serialization.blocks');
+const variables = goog.require('Blockly.serialization.variables');
+ 
+
+/**
+ * Returns the state of the workspace as a plain JavaScript object.
+ * @param {!Blockly.Workspace} workspace The workspace to serialize.
+ * @return {!Object<string, *>} The serialized state of the workspace.
+ */
+const save = function(workspace) {
+  const state = Object.create(null);
+
+  // TODO: Switch this to use plugin serialization system (once it is built).
+  const variableState = [];
+  const vars = workspace.getAllVariables();
+  for (let i = 0; i < vars.length; i++) {
+    variableState.push(variables.save(vars[i]));
+  }
+  if (variableState.length) {
+    state['variables'] = variableState;
+  }
+
+  const blockState = [];
+  for (let block of workspace.getTopBlocks()) {
+    blockState.push(
+        blocks.save(block, {addCoordinates: true, addNextBlocks: true}));
+  }
+  if (blockState.length) {
+    // This is an object to support adding language version later.
+    state['blocks'] = {
+      'blocks': blockState
+    };
+  }
+
+  return state;
+};
+ 
+/**
+ * Loads the variable represented by the given state into the given workspace.
+ * @param {!Object<string, *>} state The state of the workspace to deserialize
+ *     into the workspace.
+ * @param {!Blockly.Workspace} workspace The workspace to add the new state to.
+ */
+const load = function(state, workspace) {
+  // TODO: Switch this to use plugin serialization system (once it is built).
+  // TODO: Add something for clearing the state before deserializing.
+
+  if (state['variables']) {
+    const variableStates = state['variables'];
+    for (let i = 0; i < variableStates.length; i++) {
+      variables.load(variableStates[i], workspace);
+    }
+  }
+
+  if (state['blocks']) {
+    const blockStates = state['blocks']['blocks'];
+    for (let i = 0; i < blockStates.length; i++) {
+      blocks.load(blockStates[i], workspace);
+    }
+  }
+};
+ 
+exports = {
+  blocks,
+  load,
+  save,
+  variables,
+};
```

Comment:
Does this look correct?

Reply:
I think this should not export variables, because we want the save and load for variables to remain internal.

Reply:
The part I am not sure about is having this file export items from a separate module. This is something we do in blockly samples (index.js usually exports everything), but I am not sure how this fits into the `goog.module` conversion in core. @rachel-fenichel any thoughts?

Reply:
Two things here:
1) Let's avoid creating non-leaf namespaces going forward. In this case, that's `Blockly.serialization` given that you are also creating `Blockly.serialization.variables`, etc.
2) What is the purpose of re-exporting these functions, given that you can require the serialization code and access it from anywhere in core?

Reply:
> Let's avoid creating non-leaf namespaces going forward. In this case, that's Blockly.serialization given that you are also creating Blockly.serialization.variables, etc.

Cool, we can totally change this around since it isn't published yet. If I'm reading this right, this relates to [3.3.1](https://google.github.io/styleguide/jsguide.html#naming-hierarchy) in the styleguide? So instead of `Blockly.serialization` and `Blockly.serialization.variables`, we would have `Blockly.serialization.<something>` and `Blockly.serialization.variables`? I'm thinking `Blockly.serialization.workspace`.

> What is the purpose of re-exporting these functions, given that you can require the serialization code and access it from anywhere in core?

It wasn't letting me access the nested namespaces (eg `Blockly.serialization.variables`) from inside tests. But I'm thinking now that I was probably requiring them in requires.js incorrectly. I'll investigate that when I switch over the top-level namespace.